runtime/common/src/precompile/incentives.rs

```diff
@@ -0,0 +1,690 @@
+				let who = input.account_id_at(1)?;
+				let lp_currency_id = input.currency_id_at(2)?;
+				let amount = input.balance_at(3)?;
+
+				<module_incentives::Pallet<Runtime> as IncentivesManager<
+					Runtime::AccountId,
+					Balance,
+					CurrencyId,
+					PoolId,
+				>>::withdraw_dex_share(&who, lp_currency_id, amount)
+				.map_err(|e| PrecompileFailure::Revert {
+					exit_status: ExitRevert::Reverted,
+					output: Into::<&str>::into(e).as_bytes().to_vec(),
+					cost: target_gas_limit(target_gas).unwrap_or_default(),
+				})?;
+
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: vec![],
+					logs: Default::default(),
+				})
+			}
+			Action::ClaimRewards => {
+				let who = input.account_id_at(1)?;
+				let pool = input.u32_at(2)?;
+				let pool_currency_id = input.currency_id_at(3)?;
+				let pool_id = init_pool_id(pool, pool_currency_id, target_gas)?;
+
+				<module_incentives::Pallet<Runtime> as IncentivesManager<
+					Runtime::AccountId,
+					Balance,
+					CurrencyId,
+					PoolId,
+				>>::claim_rewards(who, pool_id)
+				.map_err(|e| PrecompileFailure::Revert {
+					exit_status: ExitRevert::Reverted,
+					output: Into::<&str>::into(e).as_bytes().to_vec(),
+					cost: target_gas_limit(target_gas).unwrap_or_default(),
+				})?;
+
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: vec![],
+					logs: Default::default(),
+				})
+			}
+			Action::GetClaimRewardDeductionRate => {
+				let pool = input.u32_at(1)?;
+				let pool_currency_id = input.currency_id_at(2)?;
+				let pool_id = init_pool_id(pool, pool_currency_id, target_gas)?;
+
+				let value = <module_incentives::Pallet<Runtime> as IncentivesManager<
+					Runtime::AccountId,
+					Balance,
+					CurrencyId,
+					PoolId,
+				>>::get_claim_reward_deduction_rate(pool_id);
+
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: Output::default().encode_u128(value.into_inner()),
+					logs: Default::default(),
+				})
+			}
+			Action::GetPendingRewards => {
+				// solidity abi encode array will add an offset at input[1]
+				let pool = input.u32_at(2)?;
+				let pool_currency_id = input.currency_id_at(3)?;
+				let pool_id = init_pool_id(pool, pool_currency_id, target_gas)?;
+				let who = input.account_id_at(4)?;
+				let reward_currency_ids_len = input.u32_at(5)?;
+				let mut reward_currency_ids = vec![];
+				for i in 0..reward_currency_ids_len {
+					reward_currency_ids.push(input.currency_id_at((6 + i) as usize)?);
+				}
+
+				let value = <module_incentives::Pallet<Runtime> as IncentivesManager<
+					Runtime::AccountId,
+					Balance,
+					CurrencyId,
+					PoolId,
+				>>::get_pending_rewards(pool_id, who, reward_currency_ids);
+
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: Output::default().encode_u128_array(value),
```

Comment:
value is not a fixed array but `encode_u128_array` will encode a fixed array without any size or offset which will be impossible to decode to the correct length. So `encode_u128_array` need to be updated to handle dynamic arrays. Same story for `encode_address_array` which is used by stableAsset precompile.

Reply:
I've opened a PR #2196